autodrain.go

```diff
@@ -0,0 +1,185 @@
+package main
+
+// nolint:gosec
+import (
+	"bytes"
+	"crypto/md5"
+	"encoding/binary"
+	"time"
+
+	"github.com/getsentry/raven-go"
+)
+
+func (s *Server) autoDrain() {
+	defer close(s.autoDrainStopped)
+
+	// Drainer is written for "efes drain" command but can be reused here because we need same functionality here.
+	d, err := NewDrainer(s.config)
+	if err != nil {
+		s.log.Errorln("Error while initializing auto-drain:", err)
+		raven.CaptureError(err, nil)
+		return
+	}
+	defer d.Shutdown() // nolint:errcheck
+
+	// Check to see if auto-drain needs to be started every minute.
+	ticker := time.NewTicker(time.Minute)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ticker.C:
+			runAutoDrain, status := s.shouldRunAutoDrain()
+			if !runAutoDrain {
+				continue
+			}
+			s.log.Info("Auto-drain has started.")
+
+			// Set drainer target to devices with usage below total average + threshold.
+			// This is not need to be checked on every file moved because we don't want to create extra
+			// load on tracker. Beside, it is not an information that changes frequently and auto-drainer
+			// will run for a limited period of time.
+			d.Dest = s.filterDevicesForAutoDrain(status)
+			if len(d.Dest) == 0 {
+				s.log.Warning("No devices available as auto-drain target")
+				continue
+			}
+
+			// Keep last fid to not selecting it again.
+			// Some files may not be movable due to a permission issue, etc.
+			var lastFid int64
+
+			for ok := true; ok; ok = s.shouldContinueAutoDrain(status.totalUse) {
+				fid, err := s.autoDrainGetNextFid(lastFid)
+				if err != nil {
+					s.log.Errorln("Error while getting next fid for auto-drain operation:", err)
+					raven.CaptureError(err, nil)
+					continue
+				}
+				lastFid = fid
+				err = d.moveFile(fid)
+				if err != nil {
+					s.log.Errorln("Error while auto-drain is moving a file:", err)
+					raven.CaptureError(err, nil)
+					continue
+				}
+			}
+			s.log.Info("Auto-drain has stopped. Waiting for next run period.")
+		case <-s.shutdown:
+			return
+		}
+	}
+}
+
+// isMyAutoDrainPeriod is here for limiting the number of parallel auto-drain operations to some portion of the cluster.
+// All servers call this method at the same time to see if they are allowed to run auto-drain for current period.
+// Period duration can be configured with Config.AutoDrainRunPeriod option.
+// To tweak the number of devices, you can use Config.AutoDrainDeviceRatio option.
+// If AutoDrainDeviceRatio is 8, only 1/8 of the total servers run auto-drain in parallel.
+func (s *Server) isMyAutoDrainPeriod(now time.Time) bool {
```

Comment:
We pass the same param at each places where we call `isMyAutoDrainPeriod` function.  So, maybe we should not pass the param and define `now time.Time` inside the function?

Reply:
I take current time as argument to those methods because I wanted them to be testable. I haven't written any tests though :)